bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/overlay/SearchHistoryMenu.java

```diff
@@ -0,0 +1,139 @@
+/*******************************************************************************
+ * Copyright (c) 2024 Vector Informatik GmbH and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Vector Informatik GmbH - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.ui.internal.findandreplace.overlay;
+
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.events.KeyListener;
+import org.eclipse.swt.events.MouseListener;
+import org.eclipse.swt.events.SelectionListener;
+import org.eclipse.swt.events.ShellAdapter;
+import org.eclipse.swt.events.ShellEvent;
+import org.eclipse.swt.events.ShellListener;
+import org.eclipse.swt.graphics.Point;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Control;
+import org.eclipse.swt.widgets.Shell;
+import org.eclipse.swt.widgets.Table;
+import org.eclipse.swt.widgets.TableColumn;
+import org.eclipse.swt.widgets.TableItem;
+
+import org.eclipse.jface.dialogs.Dialog;
+import org.eclipse.jface.layout.GridDataFactory;
+
+import org.eclipse.ui.internal.findandreplace.FindReplaceMessages;
+import org.eclipse.ui.internal.findandreplace.HistoryStore;
+
+/**
+ * Menu dropdown for the search history in the find/replace overlay
+ */
+class SearchHistoryMenu extends Dialog {
+	private final HistoryItemSelectedListener selectionListener;
+	private final HistoryStore history;
+	private final ShellListener shellFocusListener = new ShellAdapter() {
+		@Override
+		public void shellDeactivated(ShellEvent e) {
+			if (!getShell().isDisposed()) {
+				getShell().getDisplay().asyncExec(SearchHistoryMenu.this::close);
+			}
+		}
+	};
+	private Point location;
+	private int width;
+	private Table table;
+	private TableColumn column;
+
+	public SearchHistoryMenu(Shell parent, HistoryStore history,
+			HistoryItemSelectedListener menuItemSelectionListener) {
```

Comment:
It would be sufficient to use a general `Consumer<String>` instead of introducing a new (functional) interface here. If not, the listener should at least be declared as a functional interface, so that a lambda instead of an anonymous class can be passed as an argument to this constructor.